### PR TITLE
Improve accounting scripts

### PR DESCRIPTION
**Pull request by apdibbo**
Convert to Python3

Use dedicated accounting database user

Improvements to the accounting suite

Add accounting for other components

### CODE REVIEW

**Review by DavidFair — CHANGES_REQUESTED** [review]
Partial review

File: OpenStack-accounting/usr/local/sbin/manila_extract_accounting.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import accountinglib
+
+import sys
+import time
+import datetime
+
+
+nowtime = time.localtime()
+
+print("Cinder Accounting run start")
```
Comment (meoflynn) [inline OpenStack-accounting/usr/local/sbin/manila_extract_accounting.py:11]: This should be `"Manila Accounting run start"`

File: OpenStack-accounting/usr/local/sbin/nova_extract_accounting.py [hunk 2]
```diff
@@ -1,64 +1,25 @@
...
-   connectionstring = parser.get('database','connection')
-except:
-   raise RuntimeError('Unable to read from config file')
-   sys.exit(1)
-
+print("Nova Accounting run start")
 starttime=sys.argv[1]
-print(starttime)
+print("Start Time = " + starttime)
 endtime=sys.argv[2]
```
Comment (meoflynn) [inline OpenStack-accounting/usr/local/sbin/nova_extract_accounting.py:14]: What is `sys.argv` ?

File: OpenStack-accounting/sql/nova_get_accounting_data.sql [hunk 3]
```diff
@@ -1,4 +1,4 @@
-CREATE PROCEDURE `get_accounting_data`(IN starttime datetime, IN endtime datetime)
+CREATE DEFINER=`accounting-db`@`host-172-16-101-160.nubes.stfc.ac.uk` PROCEDURE `get_accounting_data`(IN starttime datetime, IN endtime datetime)
```
Comment (meoflynn) [inline OpenStack-accounting/sql/nova_get_accounting_data.sql:1]: Is the `CREATE DEFINER` needed for getting accounting data from the other OpenStack components as well?

File: OpenStack-accounting/readme.md
Comment (meoflynn) [inline OpenStack-accounting/readme.md:1]: Could we also add a `requirements.txt` file with the required python pages in the `OpenStack-accounting` folder?

File: OpenStack-accounting/usr/local/sbin/accountinglib.py
Comment (meoflynn) [inline OpenStack-accounting/usr/local/sbin/accountinglib.py:1]: Could you add docstrings to describe the methods in this file?

File: OpenStack-accounting/sql/cinder_get_accounting_data.sql
Comment (meoflynn) [inline OpenStack-accounting/sql/cinder_get_accounting_data.sql:1]: Could you add a block comment describing what the procedure does?

File: OpenStack-accounting/sql/glance_get_accounting_data.sql
Comment (meoflynn) [inline OpenStack-accounting/sql/glance_get_accounting_data.sql:1]: Could you add a block comment describing what the procedure does?

File: OpenStack-accounting/sql/manila_get_accounting_data.sql
Comment (meoflynn) [inline OpenStack-accounting/sql/manila_get_accounting_data.sql:1]: Could you add a block comment describing what the procedure does?

**Review by meoflynn — CHANGES_REQUESTED** [review]

File: OpenStack-accounting/sql/nova_get_accounting_data.sql (hunk 3 above)
Comment (apdibbo) [reply OpenStack-accounting/sql/nova_get_accounting_data.sql:1]: It isnt and is something added in by the mysql workbench that I missed removing from this one.

File: OpenStack-accounting/readme.md
Comment (apdibbo) [reply OpenStack-accounting/readme.md:1]: Done

File: OpenStack-accounting/sql/nova_get_accounting_data.sql (hunk 3 above)
Comment (apdibbo) [reply OpenStack-accounting/sql/nova_get_accounting_data.sql]: Fixed

File: OpenStack-accounting/usr/local/sbin/manila_extract_accounting.py (hunk 1 above)
Comment (apdibbo) [reply OpenStack-accounting/usr/local/sbin/manila_extract_accounting.py]: Fixed

File: OpenStack-accounting/readme.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# TheCount
+
+
+## Prerequisites
+
+The following packages are required:
+ - python-ConfigParser
```
[the remark this answers is not in the document]
Comment (apdibbo) [reply OpenStack-accounting/readme.md]: Done

File: OpenStack-accounting/usr/local/sbin/accountinglib.py
Comment (apdibbo) [reply OpenStack-accounting/usr/local/sbin/accountinglib.py:1]: Yes... *googles*

File: OpenStack-accounting/usr/local/sbin/nova_extract_accounting.py (hunk 2 above)
Comment (anish-mudaraddi) [reply OpenStack-accounting/usr/local/sbin/nova_extract_accounting.py:14]: In this case you only have 2 arguments so using `sys.argv` is fine 
Ideally using `argparse`  module you can write a more user-friendly interface - for example you specify a `--help` command.

I recommend if you're going to stick with `sys.argv` - you should raise an error if you don't receive both values and specify what the program wants. You may also want to check the inputs and make sure they are in the right format

File: OpenStack-accounting/usr/local/sbin/cinder_extract_accounting.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        if "rally" in result["Project"] :
+            department="STFC Cloud"
+        else:
+            if "default" in result["Department"]:
+                department=result["Project"]
+            else:
+                department=result["Department"]
+
+    except:
+        department="UNKNOWN"
```
Comment (anish-mudaraddi) [inline OpenStack-accounting/usr/local/sbin/cinder_extract_accounting.py:35]:
```suggestion
if "rally" in result.get("Project", ""):
    department = "STFC Cloud"
elif "default" in result.get("Department", ""):
    department = result.get("Project", "")
else:
    department = result.get("Department", "UNKNOWN")
```
try catch here is not necessary and could lead to bugs in the python code not being caught in the code block

File: OpenStack-accounting/usr/local/sbin/cinder_extract_accounting.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+
+
+    datastring += " "+str(int(endtimestamp))
+    datastring += "\n"
+
+r = accountinglib.send_to_influx(datastring)
+
+print(r.text)
+print(r)
```
Comment (anish-mudaraddi) [inline OpenStack-accounting/usr/local/sbin/cinder_extract_accounting.py:56]: this should be encased in 
`if __name__ == "__main__":`
or better yet, a `main` function which is run like so:

```
if __name__ == "__main__":
    main()
```

File: OpenStack-accounting/usr/local/sbin/glance_extract_accounting.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import accountinglib
+
+import sys
+import time
+import datetime
+
+
+nowtime = time.localtime()
+
+print("Glance Accounting run start")
```
Comment (anish-mudaraddi) [inline OpenStack-accounting/usr/local/sbin/glance_extract_accounting.py:11]: These print statements should really be logs: i.e. `logger.debug()` 
https://www.geeksforgeeks.org/logging-in-python/

File: OpenStack-accounting/usr/local/sbin/glance_extract_accounting.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    try:
+        if "rally" in result["Project"] :
+            department="STFC Cloud"
+        else:
+            if "efault" in result["Department"]:
+                department=result["Project"]
+            else:
+                department=result["Department"]
+
+    except:
```
Comment (anish-mudaraddi) [inline OpenStack-accounting/usr/local/sbin/glance_extract_accounting.py:34]: Suggestion: you should really specify the exception you're looking for 

If you're trying to catch any exception that can be raised - you should at least log what the exception was so we can track it
```
except Exceptions as exp: 
    print("An exception was raised: ", str(exp)
````

File: OpenStack-accounting/usr/local/sbin/accountinglib.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+         database = influx_parser.get('db', 'database')
+         username = influx_parser.get('auth', 'username')
+         password = influx_parser.get('auth', 'password')
+         instance = influx_parser.get('cloud','instance')
+     except:
+         print('Unable to read from influx config file')
+         sys.exit(1)
+     finaldatastring = datastring.replace("Accounting,","Accounting,instance="+instance+",")
+     print(finaldatastring)
+     url = 'http://'+host+'/write?db='+database +'&precision=s'
```
Comment (anish-mudaraddi) [inline OpenStack-accounting/usr/local/sbin/accountinglib.py:35]: you can use f-strings to make this clearer
```suggestion
     url = f"http://{host}/write?db={database}&precision=s"
```

**Review by anish-mudaraddi — CHANGES_REQUESTED** [review]
I've done a first-pass - I've raised some suggestions - these will apply to all scripts - I've just chosen a few to comment on

You should also add unit-tests to these scripts - though we can leave this for another PR :smile:

File: OpenStack-accounting/usr/local/sbin/cinder_extract_accounting.py (hunk 6 above)
Comment (anish-mudaraddi) [reply OpenStack-accounting/usr/local/sbin/cinder_extract_accounting.py:56]: you can parse `sys.argv` arguments in `if __name__ == "__main__":` and pass them in as parameters to `main()`.

File: OpenStack-accounting/sql/cinder_get_accounting_data.sql
Comment (apdibbo) [reply OpenStack-accounting/sql/cinder_get_accounting_data.sql:1]: Done

File: OpenStack-accounting/sql/glance_get_accounting_data.sql
Comment (apdibbo) [reply OpenStack-accounting/sql/glance_get_accounting_data.sql:1]: Done

File: OpenStack-accounting/sql/manila_get_accounting_data.sql
Comment (apdibbo) [reply OpenStack-accounting/sql/manila_get_accounting_data.sql:1]: Done

**Review by meoflynn — CHANGES_REQUESTED** [review]

File: OpenStack-accounting/usr/local/sbin/accountinglib.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+import json
+import requests
+import sys
+import sqlalchemy
+import logging
+from sqlalchemy.sql import select
+from sqlalchemy.orm import sessionmaker
+import configparser
+
+def get_logger(component):
```
[the remark this answers is not in the document]
Comment (apdibbo) [reply OpenStack-accounting/usr/local/sbin/accountinglib.py:14]: Done

**Comment by apdibbo** [comment]
I think this is ready now

File: OpenStack-accounting/sql/nova_get_accounting_data.sql [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            IF(i.created_at > starttime /* Captures VMs which were created during the period and deleted outside the period */
+                    AND (i.deleted_at >= endtime
+                    OR ISNULL(i.deleted_at)),
+                TIMESTAMPDIFF(SECOND,
+                    i.created_at,
+                    endtime),
+                TIMESTAMPDIFF(SECOND,
+                    i.created_at,
+                    i.deleted_at))))) AS VM_Seconds, /* Generates a count of seconds VMs were running */
+    'something' AS testingstuff,
```
Comment (meoflynn) [inline OpenStack-accounting/sql/nova_get_accounting_data.sql:33]: Is this for testing the procedure only? I think it should be removed from the query

File: OpenStack-accounting/sql/nova_get_accounting_data.sql [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+                IF(i.created_at > starttime /* Captures VMs which were created during the period and deleted outside the period */
+                        AND (i.deleted_at >= endtime
+                        OR ISNULL(i.deleted_at)),
+                    TIMESTAMPDIFF(SECOND,
+                        i.created_at,
+                        endtime),
+                    TIMESTAMPDIFF(SECOND,
+                        i.created_at,
+                        i.deleted_at))))) AS VM_Seconds, /* Generates a count of seconds VMs were running */
+        'something' AS testingstuff,
```
Comment (meoflynn) [inline OpenStack-accounting/sql/nova_get_accounting_data.sql:110]: Is this for testing the procedure only? I think it should be removed from the query

**Review by meoflynn — CHANGES_REQUESTED** [review]

File: OpenStack-accounting/sql/nova_get_accounting_data.sql (hunk 12 above)
Comment (apdibbo) [reply OpenStack-accounting/sql/nova_get_accounting_data.sql]: Yes it is, removed now

File: OpenStack-accounting/sql/nova_get_accounting_data.sql (hunk 11 above)
Comment (apdibbo) [reply OpenStack-accounting/sql/nova_get_accounting_data.sql]: Yes it is, removed now

File: OpenStack-accounting/sql/cinder_get_accounting_data.sql [hunk 13]
```diff
@@ -0,0 +1,44 @@
+CREATE PROCEDURE `get_accounting_data`(IN starttime datetime, IN endtime datetime)
+BEGIN
+/*
+   This procedure generates accounting data for cinder
+*/
+SELECT
+    IFNULL(v.availability_zone, 'nova') AS AvailabilityZone,
```
Comment (meoflynn) [inline OpenStack-accounting/sql/cinder_get_accounting_data.sql:7]: Just to double check, here is it only selecting `nova` as the availability zone if there isn't an availability zone for a given volume (e.g. volume doesn't have a ceph availability zone and is just `null`)?

File: OpenStack-accounting/sql/glance_get_accounting_data.sql [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+/*
+   This procedure generates accounting data for glance
+*/
+SELECT
+    p.name AS Project,
+    pp.name AS Department,
+    COUNT(g.id) AS Images,
+    ip.value as GlanceType,
+    if(il.value like "%rbd%", "RBD" ,if(il.value like "%swift%","OBJECT","UNKNONWN")) as StorageBackend,
+    @VolumeSeconds:=SUM(IF(g.created_at <= starttime /* Captures Images which were created outside of the period deleted out of the period */
```
Comment (meoflynn) [inline OpenStack-accounting/sql/glance_get_accounting_data.sql:12]: Should this be `@ImageSeconds` instead of `@VolumeSeconds`?

**Review by meoflynn — CHANGES_REQUESTED** [review]
Looks good to me, I just had a question about the glance accounting query

File: OpenStack-accounting/sql/cinder_get_accounting_data.sql (hunk 13 above)
Comment (apdibbo) [reply OpenStack-accounting/sql/cinder_get_accounting_data.sql:7]: Yes, that is what that's doing

File: OpenStack-accounting/sql/glance_get_accounting_data.sql (hunk 14 above)
Comment (apdibbo) [reply OpenStack-accounting/sql/glance_get_accounting_data.sql]: Yes it should. Fixed

**Review by meoflynn — APPROVED** [review]

**Review by DavidFair — DISMISSED** [review]

**Review by anish-mudaraddi — APPROVED** [review]

**Review by DavidFair — APPROVED** [review]

**Pull request merged**